metapool/plate.py

```diff
@@ -214,3 +217,198 @@ def _plate_position(well):
+
+    for i in predictions.index:
+        # Finds the first sample predicted as non-blank.
+        if predictions.loc[i, 'predicted'] == False:    # noqa: E712
+            # Returns the concentation of the first sample predicted as
+            # non-blank. This concentration is optimal boundary between
+            # classes passed as labels.
+            return predictions.loc[i, 'concentration']
+
+
+def autopool(plate_df, method='norm', pool_failures='low', automate=True,
+             offset=0.01, min_pool=100, total_vol=100, floor_conc=10,
+             min_conc=0, total_nmol=0.0040):
+    """
+    reads a plate_df and calculates pooling volumes based on parameters.
+    :param plate_df: processing-plate Pandas dataframe.
+    :param method: 'norm' (normalized pooling) or 'evp' (equal volume pooling).
+    :param pool_failures: Pool failed samples/blanks w/either 'low' or 'high'
+                          volumes. Placing them close to the left or right
+                          edge of the pooling volume distribution,
+                          respectively.
+    :param automate: estimate optimal parameters if True.
+                     Relies on manually supplied params if Fase.
+    :param offset:
+    :param min_pool:
+    :param total_vol:
+    :param floor_conc: lowest conc. for which sample will be accurately pooled.
+    :param min_conc: min conc. for a sample to be considered for pooling.
+                     Set to 0 to pool all samples regardless.
+    :param total_nmol:
+    :return: DataFrame w/calculated pooling volumes (in nL).
+    """
+
+    if method not in ['evp', 'norm']:
+        raise ErrorMessage('method must be either "evp" or "norm".')
+
+    sample_concs = plate_df['MiniPico Library Concentration']
+
+    if method == 'evp':
+        # convert to nanoliters.
+        per_sample_vol = (total_vol / sample_concs.size) * 1000.0
+        sample_vols = np.zeros(sample_concs.shape) + per_sample_vol
+        plate_df['MiniPico Pooled Volume'] = sample_vols
+    else:
+        # set proportion of pool(frac) occupied by each sample
+        sample_fracs = np.ones(sample_concs.shape) / sample_concs.size
+
+        # get samples above threshold
+        sample_fracs_pass = sample_fracs.copy()
+        sample_fracs_pass[sample_concs <= min_conc] = 0
+
+        # renormalize to exclude lost samples
+        sample_fracs_pass *= 1 / sample_fracs_pass.sum()
+
+        if pool_failures == 'low':
+            # calculate volumetric fractions including floor val
+            sample_vols = (total_nmol * sample_fracs_pass) / sample_concs
+
+            # convert L to nL
+            sample_vols *= 10 ** 9
+            # drop volumes for samples below floor concentration to left side
+            # of distribution.
+            # offset allows to place samples below floor concentration at a
+            # relative distance from edge of distribution.
+            sample_vols[sample_concs < floor_conc] = \
```

Comment:
Also this one ...

Reply:
I refactored autopool() once more, with this issue in mind. Should be addressed now.